lib.go

```diff
@@ -51,9 +51,10 @@ func (vm *VM) Cleanup() {
 // This function stores the code for that contract only once, but it can
 // be instantiated with custom inputs in the future.
 //
-// TODO: return gas cost? Add gas limit??? there is no metering here...
-func (vm *VM) StoreCode(code WasmCode) (Checksum, error) {
-	return api.StoreCode(vm.cache, code)
+// Returns both the checksum, as well as the gas cost of compilation (in CosmWasm Gas) or an error.
+func (vm *VM) StoreCode(code WasmCode) (Checksum, uint64, error) {
```

Comment:
What we'd also need is a gas limit. If the transaction sets a small gas limit we should be able to stop with out of gas early before doing the heavy computation.

Reply:
So, if it would be above the limit, we return 0 gas usage here, right?

Reply:
The important part is to error with out of gas. Then we have two options:

1. Error with out of gas and report `compileCosts(code)` as gas used
2. Error with out of gas and report 0 as gas used

Version 2. has the problem that you get fun errors like: _Out of gas. Used: 3, limit 9000_.

Which of those two is the current behaviour in wasmd?

Reply:
Wasmd just calls the SDK `GasMeter`'s `ConsumeGas` with the calculated amount. So it's only limited by the amount left in the `GasMeter`.
After looking a bit at how wasmd handles this in `Execute`, I think version 1 is the way to go:
The gas limit passed to wasmvm is the remaining gas in the `GasMeter`.
After the wasmvm call, wasmd calls `k.consumeRuntimeGas(sdkCtx, gasUsed)`, which would then cause an out of gas error if it's more than the remaining gas. So in order to get the _Out of gas_, we actually have to return more `gasUsed` than the limit we got.